### PR TITLE
update for upstream changes to TyFnDef

### PR DESCRIPTION
**Pull request by dwrensha**
Fixes breakage from https://github.com/rust-lang/rust/pull/42417.

### CODE REVIEW

**Comment by dwrensha** [comment]
The `tests/run-pass/non_capture_closure_to_fn_ptr.rs` test still fails:

```
$ cargo run --bin miri -- tests/run-pass/non_capture_closure_to_fn_ptr.rs 
   Compiling miri v0.1.0 (file:///home/david/src/miri)
    Finished dev [unoptimized + debuginfo] target(s) in 35.92 secs
     Running `target/debug/miri tests/run-pass/non_capture_closure_to_fn_ptr.rs`
error: tried to call a function with sig extern "rust-call" fn([closure@tests/run-pass/non_capture_closure_to_fn_ptr.rs:3:20: 3:45], ()) -> <[closure@tests/run-pass/non_capture_closure_to_fn_ptr.rs:3:20: 3:45] as std::ops::FnOnce<()>>::Output through a function pointer of type fn()
 --> tests/run-pass/non_capture_closure_to_fn_ptr.rs:8:5
  |
8 |     FOO();
  |     ^^^^^
  |
note: inside call to main
 --> tests/run-pass/non_capture_closure_to_fn_ptr.rs:7:1
  |
7 | / fn main() {
8 | |     FOO();
9 | |     BAR(44, 45);
10| |     let bar: unsafe fn(i32, i32) = BAR;
... |
13| |     boo(48, 49);
14| | }
  | |_^

error: aborting due to previous error(s)

thread 'main' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:512
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by dwrensha** [comment]
I think maybe we need to update the [`check_sig_compat()` method](https://github.com/solson/miri/blob/695f9723d2046d73566a10660f411dfcedb16411/src/terminator/mod.rs#L142-L202)?

File: src/eval_context.rs [hunk 1]
```diff
@@ -1891,9 +1891,11 @@ pub fn resolve<'a, 'tcx>(
     } else {
         let item_type = def_ty(tcx, def_id, substs);
         let def = match item_type.sty {
-            ty::TyFnDef(_, _, f) if
-                f.abi() == Abi::RustIntrinsic ||
-                f.abi() == Abi::PlatformIntrinsic =>
+            ty::TyFnDef(..) if {
+                    let f = item_type.fn_sig(tcx);
```
Comment (eddyb) [inline src/eval_context.rs]: Do not call this. `tcx.fn_sig(def_id)` is less work.

**Comment by RalfJung** [comment]
> I think maybe we need to update the check_sig_compat() method?

To me this looks more like some form of normalization used to be performed, but is no longer performed... I never saw types like `<[closure@tests/run-pass/non_capture_closure_to_fn_ptr.rs:3:20: 3:45] as std::ops::FnOnce<()>>::Output` when I wrote `check_sig_compat`.

**Comment by eddyb** [comment]
Ah yeah you need to normalize yourself when working on a signature, like you would a field.

File: src/eval_context.rs (hunk 1 above)
Comment (dwrensha) [inline src/eval_context.rs]: This code is directly copied from [librustc_trans](https://github.com/rust-lang/rust/blob/dc2003b7cc1af7d7beee0b1ceec3d811c28cb35e/src/librustc_trans/monomorphize.rs#L168-L172). Should it also be changed there?

Comment (eddyb) [inline src/eval_context.rs]: Ah I see, just leave it alone then. Minor inefficiency but the `DefId` might not match.

**Comment by dwrensha** [comment]
OK. Fixed by adding a call to `self.tcx.normalize_associated_type()`.

**Pull request merged**